Kmesh daemon crashed

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!
-->

**What happened**:

I tried to run e2e tests in ipv6 env:

```bash
root@kurator-linux-0002:~/kmesh# pwd
/root/kmesh
root@kurator-linux-0002:~/kmesh# ./test/e2e/run_test.sh --ipv6 
```

But the Kmesh daemon pod in worker node keeps crashing:

```bash
root@kurator-linux-0002:~/kmesh# kubectl get pods -n kmesh-system  -owide  
NAME          READY   STATUS             RESTARTS        AGE   IP                 NODE                          NOMINATED NODE   READINESS GATES
kmesh-r6fmh   0/1     CrashLoopBackOff   6 (3m51s ago)   10m   fd00:10:244:1::4   kmesh-testing-worker          <none>           <none>
kmesh-slzbp   1/1     Running            0               10m   fd00:10:244::6     kmesh-testing-control-plane   <none>           <none>
```

The log of crashing Kmesh daemon pod:

```bash
time="2024-08-08T07:01:51Z" level=warning msg="get workload from ip [253 0 0 16 2 68 0 1 0 0 0 0 0 0 0 1] FAILED" subsys=pkg/auth
time="2024-08-08T07:01:51Z" level=warning msg="Auth denied for connection: [253 0 0 16 2 68 0 1 0 0 0 0 0 0 0 5] because destination workload not found" subsys=pkg/auth
time="2024-08-08T07:01:51Z" level=info msg="Auth denied for connection: {srcIdentity:{trustDomain: namespace: serviceAccount:} dstNetwork: srcIp:[253 0 0 16 2 68 0 1 0 0 0 0 0 0 0 1] dstIp:[253 0 0 16 2 68 0 1 0 0 0 0 0 0 0 5] dstPort:8080}" subsys=pkg/auth
time="2024-08-08T07:01:51Z" level=warning msg="get workload from ip [253 0 0 16 2 68 0 1 0 0 0 0 0 0 0 1] FAILED" subsys=pkg/auth
time="2024-08-08T07:01:51Z" level=warning msg="Auth denied for connection: [253 0 0 16 2 68 0 1 0 0 0 0 0 0 0 6] because destination workload not found" subsys=pkg/auth
time="2024-08-08T07:01:51Z" level=info msg="Auth denied for connection: {srcIdentity:{trustDomain: namespace: serviceAccount:} dstNetwork: srcIp:[253 0 0 16 2 68 0 1 0 0 0 0 0 0 0 1] dstIp:[253 0 0 16 2 68 0 1 0 0 0 0 0 0 0 6] dstPort:3333}" subsys=pkg/auth
time="2024-08-08T07:01:51Z" level=warning msg="get workload from ip [253 0 0 16 2 68 0 1 0 0 0 0 0 0 0 1] FAILED" subsys=pkg/auth
time="2024-08-08T07:01:51Z" level=warning msg="Auth denied for connection: [253 0 0 16 2 68 0 1 0 0 0 0 0 0 0 6] because destination workload not found" subsys=pkg/auth
time="2024-08-08T07:01:51Z" level=info msg="Auth denied for connection: {srcIdentity:{trustDomain: namespace: serviceAccount:} dstNetwork: srcIp:[253 0 0 16 2 68 0 1 0 0 0 0 0 0 0 1] dstIp:[253 0 0 16 2 68 0 1 0 0 0 0 0 0 0 6] dstPort:8080}" subsys=pkg/auth
time="2024-08-08T07:01:51Z" level=warning msg="exceed the max port count,current only support maximum of 10 ports" subsys=workload_controller
time="2024-08-08T07:01:51Z" level=warning msg="exceed the max port count,current only support maximum of 10 ports" subsys=workload_controller
time="2024-08-08T07:01:51Z" level=info msg="reload workload config from last epoch" subsys=workload_controller
time="2024-08-08T07:01:51Z" level=info msg="echo-1-86160/enrolled-to-kmesh-v1-76f577cfc6-n4r9p: enable Kmesh manage" subsys=manage_controller
time="2024-08-08T07:01:51Z" level=info msg="echo-1-86160/enrolled-to-kmesh-v2-785d4f7df4-7xccg: enable Kmesh manage" subsys=manage_controller
time="2024-08-08T07:01:51Z" level=info msg="echo-1-86160/service-with-waypoint-at-service-granularity-v1-5dbf88d97bfc2m8: enable Kmesh manage" subsys=manage_controller
time="2024-08-08T07:01:51Z" level=info msg="echo-1-86160/service-with-waypoint-at-service-granularity-v2-5844948d94blx2f: enable Kmesh manage" subsys=manage_controller
time="2024-08-08T07:01:51Z" level=info msg="add annotation for pod echo-1-86160/enrolled-to-kmesh-v1-76f577cfc6-n4r9p" subsys=manage_controller
time="2024-08-08T07:01:51Z" level=info msg="add annotation for pod echo-1-86160/enrolled-to-kmesh-v2-785d4f7df4-7xccg" subsys=manage_controller
time="2024-08-08T07:01:51Z" level=info msg="add annotation for pod echo-1-86160/service-with-waypoint-at-service-granularity-v1-5dbf88d97bfc2m8" subsys=manage_controller
time="2024-08-08T07:01:51Z" level=info msg="add annotation for pod echo-1-86160/service-with-waypoint-at-service-granularity-v2-5844948d94blx2f" subsys=manage_controller
time="2024-08-08T07:01:51Z" level=warning msg="exceed the max port count,current only support maximum of 10 ports" subsys=workload_controller
time="2024-08-08T07:01:51Z" level=warning msg="exceed the max port count,current only support maximum of 10 ports" subsys=workload_controller
time="2024-08-08T07:01:51Z" level=info msg="Copied /usr/bin/kmesh-cni to /opt/cni/bin." subsys="cni installer"
time="2024-08-08T07:01:51Z" level=info msg="cni config file: /etc/cni/net.d/10-kindnet.conflist" subsys="cni installer"
time="2024-08-08T07:01:51Z" level=info msg="kmesh-cni was installed but not cleaned up, but we would overwrite it" subsys="cni installer"
time="2024-08-08T07:01:51Z" level=info msg="command Start cni successful" subsys=manager
time="2024-08-08T07:01:52Z" level=warning msg="get workload from ip [253 0 0 16 2 68 0 1 0 0 0 0 0 0 0 1] FAILED" subsys=pkg/telemetry
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x100 pc=0x555ea1568c0f]

goroutine 16 [running]:
kmesh.net/kmesh/pkg/controller/telemetry.buildWorkloadMetric(_, _)
        /kmesh/pkg/controller/telemetry/metric.go:323 +0x16f
kmesh.net/kmesh/pkg/controller/telemetry.(*MetricController).buildWorkloadMetric(_, _)
        /kmesh/pkg/controller/telemetry/metric.go:268 +0x285
kmesh.net/kmesh/pkg/controller/telemetry.(*MetricController).Run(0xc00071cc60, {0x555ea2c1f5e0, 0xc000687e00}, 0xc0006aacd0?)
        /kmesh/pkg/controller/telemetry/metric.go:199 +0x545
created by kmesh.net/kmesh/pkg/controller/workload.(*Controller).Run in goroutine 1
        /kmesh/pkg/controller/workload/workload_controller.go:58 +0x14d
kmesh exit
```

**What you expected to happen**:

All components are functioning properly

**How to reproduce it (as minimally and precisely as possible)**:

As above.

**Anything else we need to know?**:

**Environment**:
- Kmesh version:
- Others:

--- a/pkg/controller/telemetry/metric.go
+++ b/pkg/controller/telemetry/metric.go
@@ -304,90 +304,89 @@ func (m *MetricController) getWorkloadByAddress(address []byte) (*workloadapi.Wo
 }
 
 func buildWorkloadMetric(dstWorkload, srcWorkload *workloadapi.Workload) workloadMetricLabels {
-	if dstWorkload == nil && srcWorkload == nil {
-		return workloadMetricLabels{}
-	}
-
 	trafficLabels := workloadMetricLabels{}
 
-	trafficLabels.destinationPodNamespace = dstWorkload.Namespace
-	trafficLabels.destinationPodName = dstWorkload.Name
-	trafficLabels.destinationWorkload = dstWorkload.WorkloadName
-	trafficLabels.destinationCanonicalService = dstWorkload.CanonicalName
-	trafficLabels.destinationCanonicalRevision = dstWorkload.CanonicalRevision
-	trafficLabels.destinationWorkloadNamespace = dstWorkload.Namespace
-	trafficLabels.destinationApp = dstWorkload.CanonicalName
-	trafficLabels.destinationVersion = dstWorkload.CanonicalRevision
-	trafficLabels.destinationCluster = dstWorkload.ClusterId
+	if dstWorkload != nil {
+		trafficLabels.destinationPodNamespace = dstWorkload.Namespace
+		trafficLabels.destinationPodName = dstWorkload.Name
+		trafficLabels.destinationWorkload = dstWorkload.WorkloadName
+		trafficLabels.destinationCanonicalService = dstWorkload.CanonicalName
+		trafficLabels.destinationCanonicalRevision = dstWorkload.CanonicalRevision
+		trafficLabels.destinationWorkloadNamespace = dstWorkload.Namespace
+		trafficLabels.destinationApp = dstWorkload.CanonicalName
+		trafficLabels.destinationVersion = dstWorkload.CanonicalRevision
+		trafficLabels.destinationCluster = dstWorkload.ClusterId
+		trafficLabels.destinationPrincipal = buildPrincipal(dstWorkload)
+	}
 
-	trafficLabels.sourceWorkload = srcWorkload.WorkloadName
-	trafficLabels.sourceCanonicalService = srcWorkload.CanonicalName
-	trafficLabels.sourceCanonicalRevision = srcWorkload.CanonicalRevision
-	trafficLabels.sourceWorkloadNamespace = srcWorkload.Namespace
-	trafficLabels.sourceApp = srcWorkload.CanonicalName
-	trafficLabels.sourceVersion = srcWorkload.CanonicalRevision
-	trafficLabels.sourceCluster = srcWorkload.ClusterId
-
-	trafficLabels.destinationPrincipal = buildPrincipal(dstWorkload)
-	trafficLabels.sourcePrincipal = buildPrincipal(srcWorkload)
+	if srcWorkload != nil {
+		trafficLabels.sourceWorkload = srcWorkload.WorkloadName
+		trafficLabels.sourceCanonicalService = srcWorkload.CanonicalName
+		trafficLabels.sourceCanonicalRevision = srcWorkload.CanonicalRevision
+		trafficLabels.sourceWorkloadNamespace = srcWorkload.Namespace
+		trafficLabels.sourceApp = srcWorkload.CanonicalName
+		trafficLabels.sourceVersion = srcWorkload.CanonicalRevision
+		trafficLabels.sourceCluster = srcWorkload.ClusterId
+		trafficLabels.sourcePrincipal = buildPrincipal(srcWorkload)
+	}
 
 	return trafficLabels
 }
 
 func buildServiceMetric(dstWorkload, srcWorkload *workloadapi.Workload, dstPort uint16) serviceMetricLabels {
-	if dstWorkload == nil && srcWorkload == nil {
-		return serviceMetricLabels{}
-	}
-
 	trafficLabels := serviceMetricLabels{}
 
-	namespacedhost := ""
-	for k, portList := range dstWorkload.Services {
-		for _, port := range portList.Ports {
-			if port.TargetPort == uint32(dstPort) {
-				namespacedhost = k
+	if dstWorkload != nil {
+		namespacedhost := ""
+		for k, portList := range dstWorkload.Services {
+			for _, port := range portList.Ports {
+				if port.TargetPort == uint32(dstPort) {
+					namespacedhost = k
+					break
+				}
+			}
+			if namespacedhost != "" {
 				break
 			}
 		}
-		if namespacedhost != "" {
-			break
+		if namespacedhost == "" {
+			log.Infof("can't find service correspond workload: %s", dstWorkload.Name)
 		}
+
+		svcHost := ""
+		svcNamespace := ""
+		if len(strings.Split(namespacedhost, "/")) != 2 {
+			log.Info("get destination service host failed")
+		} else {
+			svcNamespace = strings.Split(namespacedhost, "/")[0]
+			svcHost = strings.Split(namespacedhost, "/")[1]
+		}
+
+		trafficLabels.destinationService = svcHost
+		trafficLabels.destinationServiceNamespace = svcNamespace
+		trafficLabels.destinationServiceName = svcHost
+
+		trafficLabels.destinationWorkload = dstWorkload.WorkloadName
+		trafficLabels.destinationCanonicalService = dstWorkload.CanonicalName
+		trafficLabels.destinationCanonicalRevision = dstWorkload.CanonicalRevision
+		trafficLabels.destinationWorkloadNamespace = dstWorkload.Namespace
+		trafficLabels.destinationApp = dstWorkload.CanonicalName
+		trafficLabels.destinationVersion = dstWorkload.CanonicalRevision
+		trafficLabels.destinationCluster = dstWorkload.ClusterId
+		trafficLabels.destinationPrincipal = buildPrincipal(dstWorkload)
+		trafficLabels.destinationPrincipal = buildPrincipal(dstWorkload)
 	}
-	if namespacedhost == "" {
-		log.Infof("can't find service correspond workload: %s", dstWorkload.Name)
+
+	if srcWorkload != nil {
+		trafficLabels.sourceWorkload = srcWorkload.WorkloadName
+		trafficLabels.sourceCanonicalService = srcWorkload.CanonicalName
+		trafficLabels.sourceCanonicalRevision = srcWorkload.CanonicalRevision
+		trafficLabels.sourceWorkloadNamespace = srcWorkload.Namespace
+		trafficLabels.sourceApp = srcWorkload.CanonicalName
+		trafficLabels.sourceVersion = srcWorkload.CanonicalRevision
+		trafficLabels.sourceCluster = srcWorkload.ClusterId
+		trafficLabels.sourcePrincipal = buildPrincipal(srcWorkload)
 	}
-
-	svcHost := ""
-	svcNamespace := ""
-	if len(strings.Split(namespacedhost, "/")) != 2 {
-		log.Info("get destination service host failed")
-	} else {
-		svcNamespace = strings.Split(namespacedhost, "/")[0]
-		svcHost = strings.Split(namespacedhost, "/")[1]
-	}
-
-	trafficLabels.destinationService = svcHost
-	trafficLabels.destinationServiceNamespace = svcNamespace
-	trafficLabels.destinationServiceName = svcHost
-
-	trafficLabels.destinationWorkload = dstWorkload.WorkloadName
-	trafficLabels.destinationCanonicalService = dstWorkload.CanonicalName
-	trafficLabels.destinationCanonicalRevision = dstWorkload.CanonicalRevision
-	trafficLabels.destinationWorkloadNamespace = dstWorkload.Namespace
-	trafficLabels.destinationApp = dstWorkload.CanonicalName
-	trafficLabels.destinationVersion = dstWorkload.CanonicalRevision
-	trafficLabels.destinationCluster = dstWorkload.ClusterId
-
-	trafficLabels.sourceWorkload = srcWorkload.WorkloadName
-	trafficLabels.sourceCanonicalService = srcWorkload.CanonicalName
-	trafficLabels.sourceCanonicalRevision = srcWorkload.CanonicalRevision
-	trafficLabels.sourceWorkloadNamespace = srcWorkload.Namespace
-	trafficLabels.sourceApp = srcWorkload.CanonicalName
-	trafficLabels.sourceVersion = srcWorkload.CanonicalRevision
-	trafficLabels.sourceCluster = srcWorkload.ClusterId
-
-	trafficLabels.destinationPrincipal = buildPrincipal(dstWorkload)
-	trafficLabels.sourcePrincipal = buildPrincipal(srcWorkload)
 
 	return trafficLabels
 }

--- a/pkg/controller/telemetry/metric_test.go
+++ b/pkg/controller/telemetry/metric_test.go
@@ -24,7 +24,6 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 	dto "github.com/prometheus/client_model/go"
 	"github.com/stretchr/testify/assert"
-
 	"kmesh.net/kmesh/api/v2/workloadapi"
 	"kmesh.net/kmesh/pkg/controller/workload/cache"
 )
@@ -579,6 +578,129 @@ func TestBuildServiceMetric(t *testing.T) {
 			m.workloadCache.AddWorkload(dstWorkload)
 			m.workloadCache.AddWorkload(srcWorkload)
 			got := m.buildServiceMetric(tt.args.data)
+			assert.Equal(t, tt.want, got)
+		})
+	}
+}
+
+func Test_buildServiceMetric(t *testing.T) {
+	type args struct {
+		dstWorkload *workloadapi.Workload
+		srcWorkload *workloadapi.Workload
+		dstPort     uint16
+	}
+	tests := []struct {
+		name string
+		args args
+		want serviceMetricLabels
+	}{
+		{
+			name: "normal capability test",
+			args: args{
+				dstWorkload: &workloadapi.Workload{
+					Namespace:         "kmesh-system",
+					Name:              "kmesh",
+					WorkloadName:      "kmesh-daemon",
+					CanonicalName:     "dstCanonical",
+					CanonicalRevision: "dstVersion",
+					ClusterId:         "Kubernetes",
+					TrustDomain:       "cluster.local",
+					ServiceAccount:    "default",
+					Services: map[string]*workloadapi.PortList{
+						"kmesh-system/kmesh": &workloadapi.PortList{
+							Ports: []*workloadapi.Port{
+								{
+									ServicePort: 80,
+									TargetPort:  8000,
+								},
+							},
+						},
+					},
+				},
+				srcWorkload: &workloadapi.Workload{
+					Namespace:         "kmesh-system",
+					Name:              "kmesh",
+					WorkloadName:      "kmesh-daemon",
+					CanonicalName:     "srcCanonical",
+					CanonicalRevision: "srcVersion",
+					ClusterId:         "Kubernetes",
+					TrustDomain:       "cluster.local",
+					ServiceAccount:    "default",
+				},
+				dstPort: uint16(8000),
+			},
+			want: serviceMetricLabels{
+				reporter:                     "",
+				sourceWorkload:               "kmesh-daemon",
+				sourceCanonicalService:       "srcCanonical",
+				sourceCanonicalRevision:      "srcVersion",
+				sourceWorkloadNamespace:      "kmesh-system",
+				sourcePrincipal:              "spiffe://cluster.local/ns/kmesh-system/sa/default",
+				sourceApp:                    "srcCanonical",
+				sourceVersion:                "srcVersion",
+				sourceCluster:                "Kubernetes",
+				destinationService:           "kmesh",
+				destinationServiceNamespace:  "kmesh-system",
+				destinationServiceName:       "kmesh",
+				destinationWorkload:          "kmesh-daemon",
+				destinationCanonicalService:  "dstCanonical",
+				destinationCanonicalRevision: "dstVersion",
+				destinationWorkloadNamespace: "kmesh-system",
+				destinationPrincipal:         "spiffe://cluster.local/ns/kmesh-system/sa/default",
+				destinationApp:               "dstCanonical",
+				destinationVersion:           "dstVersion",
+				destinationCluster:           "Kubernetes",
+				requestProtocol:              "",
+				responseFlags:                "",
+				connectionSecurityPolicy:     "",
+			},
+		},
+		{
+			name: "nil destination workload",
+			args: args{
+				dstWorkload: &workloadapi.Workload{},
+				srcWorkload: &workloadapi.Workload{
+					Namespace:         "kmesh-system",
+					Name:              "kmesh",
+					WorkloadName:      "kmesh-daemon",
+					CanonicalName:     "srcCanonical",
+					CanonicalRevision: "srcVersion",
+					ClusterId:         "Kubernetes",
+					TrustDomain:       "cluster.local",
+					ServiceAccount:    "default",
+				},
+				dstPort: uint16(8000),
+			},
+			want: serviceMetricLabels{
+				reporter:                     "",
+				sourceWorkload:               "kmesh-daemon",
+				sourceCanonicalService:       "srcCanonical",
+				sourceCanonicalRevision:      "srcVersion",
+				sourceWorkloadNamespace:      "kmesh-system",
+				sourcePrincipal:              "spiffe://cluster.local/ns/kmesh-system/sa/default",
+				sourceApp:                    "srcCanonical",
+				sourceVersion:                "srcVersion",
+				sourceCluster:                "Kubernetes",
+				destinationService:           "",
+				destinationServiceNamespace:  "",
+				destinationServiceName:       "",
+				destinationWorkload:          "",
+				destinationCanonicalService:  "",
+				destinationCanonicalRevision: "",
+				destinationWorkloadNamespace: "",
+				destinationPrincipal:         "-",
+				destinationApp:               "",
+				destinationVersion:           "",
+				destinationCluster:           "",
+				requestProtocol:              "",
+				responseFlags:                "",
+				connectionSecurityPolicy:     "",
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := buildServiceMetric(tt.args.dstWorkload, tt.args.srcWorkload, tt.args.dstPort)
 			assert.Equal(t, tt.want, got)
 		})
 	}

--- a/pkg/controller/telemetry/metric_test.go
+++ b/pkg/controller/telemetry/metric_test.go
@@ -24,6 +24,7 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 	dto "github.com/prometheus/client_model/go"
 	"github.com/stretchr/testify/assert"
+
 	"kmesh.net/kmesh/api/v2/workloadapi"
 	"kmesh.net/kmesh/pkg/controller/workload/cache"
 )
@@ -607,7 +608,7 @@ func Test_buildServiceMetric(t *testing.T) {
 					TrustDomain:       "cluster.local",
 					ServiceAccount:    "default",
 					Services: map[string]*workloadapi.PortList{
-						"kmesh-system/kmesh": &workloadapi.PortList{
+						"kmesh-system/kmesh": {
 							Ports: []*workloadapi.Port{
 								{
 									ServicePort: 80,

